src/tests/unittests/test_filmicrgb.c

```diff
@@ -0,0 +1,145 @@
+    // expected norms...
+    { 0.0f, 0.5f, 3.0f, 1.1f },               // max_rgb
+    { 0.0f, 0.358439f, 1.838112f, 0.244755f } // luminance
+    // TODO: luminance values are taken experimentally, just for demonstration.
+    // They should be computed by some means for testing of course!
+  };
+
+  float norm = 0.0f;
+
+  for(int v = 0; v < 2; v++)
+  {
+    printf("variant=%i\n", v);
+    for(int i = 0; i < N; i++)
+    {
+      const float *const pixel = pixels[i];
+      printf("case %i: pixel={%f, %f, %f, %f}", i, pixel[0], pixel[1], pixel[2], pixel[3]);
+      norm = get_pixel_norm(pixel, variants[v], work_profile);
+      printf(" -> norm=%f\n", norm);
+      // TODO: use assert_float_equal() - not available in Ubuntu 18.04 version
+      assert_true(norm < (exp_norms[v][i] + E));
+      assert_true(norm > (exp_norms[v][i] - E));
+    }
+  }
+#undef N
+}
+
+/*
+ * Verify the process() method, given a simple environment and some test data
+ *
+ * This is a more advanced test that covers way more than just a method.
+ * Actually, this is not really a unit test anymore. But I think it has a very
+ * nice granularity since this way we can test each module separately and feed
+ * it with simple and very small "test images". As an assertion we could have
+ * simple boundary checks or more dedicated expectations values. I also plan to
+ * add kind of a "basic set" of test inputs that each module must pass (e.g.
+ * bounded output values etc.).
+ */
+static void test_process(void **state)
+{
+  // prepare environment (TODO: is there a better way?):
+  struct dt_iop_module_t module;
+  init(&module);
+  module.commit_params = commit_params;
+  printf("module ok\n");
+  struct dt_dev_pixelpipe_t pipe;
+  struct dt_dev_pixelpipe_iop_t piece;
+  init_pipe(&module, &pipe, &piece);
+  piece.pipe = &pipe;
+  printf("pipe ok\n");
+
+  // prepare input and output buffers:
+  // TODO: of course this is just a start...
+  const float in[1][4] = { { 0.4f, 0.99f, 1.2f, 0.0f } };
+  float out[1][4] = { { 0.0f, 0.0f, 0.0f, 0.0f } };
+  dt_iop_roi_t roi_in;
+  dt_iop_roi_t roi_out;
+  roi_out.width = 1;
+  roi_out.height = 1;
+
+  // set parameters:
+  dt_iop_filmicrgb_data_t *data = (dt_iop_filmicrgb_data_t *)piece.data;
+  data->preserve_color = DT_FILMIC_METHOD_MAX_RGB;
+
+  // now, call process():
+  printf("calling process()...\n");
+  process(&module, &piece, in, out, &roi_in, &roi_out);
+
+  // assert return value:
+  // TODO: we just print here - of course, some assertion is needed :-)
+  for(size_t k = 0; k < roi_out.height * roi_out.width; k += 1)
+  {
+    float *const pix_out = out[k];
+    printf("[%f, %f, %f, %f]\n", pix_out[0], pix_out[1], pix_out[2], pix_out[3]);
+  }
+}
+
+
+/******************************************************************************
+ * MAIN FUNCTION
+ */
+int main()
+{
+  const struct CMUnitTest tests[]
+      = { cmocka_unit_test(test_sample), cmocka_unit_test(test_get_pixel_norm), cmocka_unit_test(test_process) };
```

Comment:
Please put one test per line, which is easier to read.

Reply:
This was clang-format, I will fix it.